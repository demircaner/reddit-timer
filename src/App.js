import React from 'react';
import { Switch, Route } from 'react-router-dom';
import styled, { ThemeProvider } from 'styled-components';
import Footer from './components/Footer';
import Header from './components/Header';
import HomePage from './pages/HomePage';
import SearchPage from './pages/SearchPage';
import GlobalStyle from './style/GlobalStyle';
import theme from './style/theme';

export default function App() {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      <Header />
      <ContentContainer>
        <Switch>
          <Route exact path="/">
            <HomePage />
          </Route>
          <Route path="/search/:subreddit">
            <SearchPage />
          </Route>
          <Route path="/terms">Terms page</Route>
          <Route>404 - Not Found</Route>
        </Switch>
      </ContentContainer>
      <Footer />
    </ThemeProvider>
  );
}

const ContentContainer = styled.main`
  min-height: ${(props) => `calc(100vh - ${props.theme.size.headerHeight} 
  - ${props.theme.size.footerHeight})`};
  display: flex;
  /* justify-content: center; */
  flex-direction: column;
`;
