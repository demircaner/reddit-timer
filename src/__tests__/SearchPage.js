import React from 'react';
// eslint-disable-next-line object-curly-newline
import { render, screen, within, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter, Route } from 'react-router-dom';
import App from '../App';
import { defaultSubreddit } from '../config';

function setup(initialPath = '/') {
  let history;
  const view = render(
    <MemoryRouter initialEntries={[initialPath]}>
      <App />
      <Route
        path="*"
        render={(props) => {
          history = props.history;
          return null;
        }}
      />
    </MemoryRouter>,
  );
  return { ...view, history };
}

describe('Subreddit Form', () => {
  test('Input value changes to defaultsubreddit when you navigate to search page', () => {
    setup('/search/reactjs');
    const subredditInput = screen.getByRole('textbox');
    const header = screen.getByRole('banner');
    const searchLink = within(header).getByRole('link', { name: /Search/ });

    userEvent.click(searchLink);
    expect(subredditInput.value).toBe(defaultSubreddit);
  });

  test('Updates the URL with the input value when search button is clicked', () => {
    const { history } = setup('/search/reactjs');
    const subredditInput = screen.getByLabelText('r /');
    expect(subredditInput.value).toBe('reactjs');

    userEvent.clear(subredditInput);
    userEvent.type(subredditInput, 'vuejs');
    expect(subredditInput.value).toBe('vuejs');

    const searchButton = screen.getByRole('button', {
      name: /search/i,
    });
    userEvent.click(searchButton);
    expect(history.location.pathname).toEqual('/search/vuejs');
  });

  // test('Loads top posts for the subreddit in the URL', async () => {
  //   setup('/search/javascript');
  //   const spinner = screen.getByTestId('spinner');
  //   expect(spinner).toBeInTheDocument();
  //   await waitFor(() => expect(screen.getByRole('table')).toBeInTheDocument(), {
  //     timeout: 8000,
  //   });
  //   const box = screen.getAllByRole('gridcell')[0];
  //   fireEvent.click(box);
  //   expect(box).toHaveStyle('outline: 1px solid black');
  //   expect(spinner).not.toBeInTheDocument();
  // });

  // test('renders error message', async () => {
  //   setup('/search/failing-request');
  //   await waitFor(
  //     () =>
  //       // eslint-disable-next-line implicit-arrow-linebreak
  //       expect(screen.getByText(/something went wrong/i)).toBeInTheDocument(),
  //     { timeout: 5000 },
  //   );
  // });
});

describe('heatmap', () => {
  test('loads and renders top posts for subreddit in URL', async () => {
    setup('/search/reactjs');

    const spinner = screen.getByTestId('spinner');

    await waitFor(
      () => expect(screen.getByTestId('heatmap')).toBeInTheDocument(),
      {
        timeout: 10000,
      },
    );

    // expect(await screen.findByTestId('heatmap')).toBeInTheDocument();

    expect(spinner).not.toBeInTheDocument();

    const heatmap = screen.getByTestId('heatmap');
    const cells = await within(heatmap).findAllByRole('button');
    expect(cells.length).toEqual(7 * 24);

    const numberOfPostsPerCell = cells.map((cell) => cell.innerHTML);
    expect(numberOfPostsPerCell).toMatchSnapshot();

    const timezone = screen.getByText('All times are shown in your timezone:');
    expect(within(timezone).getByText('Europe/Istanbul')).toBeInTheDocument();
  });

  test('cell highlights on click', async () => {
    setup('/search/reactjs');

    const heatmap = await screen.findByTestId('heatmap');
    const cells = await within(heatmap).findAllByRole('button');

    const cellToClick = cells[1];
    expect(cellToClick).toHaveStyle('border: none');

    userEvent.click(cellToClick);
    expect(cellToClick).toHaveStyle('border: 1px solid #1e2537');
  });

  test('renders error message', async () => {
    setup('/search/failing-request');

    expect(
      await screen.findByText(/something went wrong/i),
    ).toBeInTheDocument();
    expect(screen.queryByText('loading-spinner.svg')).not.toBeInTheDocument();
  });
});
