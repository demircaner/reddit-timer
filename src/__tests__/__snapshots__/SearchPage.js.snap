// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`heatmap cell highlights on click 1`] = `
Array [
  "0",
  "2",
  "1",
  "0",
  "3",
  "2",
  "2",
  "0",
  "1",
  "2",
  "1",
  "4",
  "3",
  "2",
  "2",
  "6",
  "3",
  "7",
  "7",
  "2",
  "4",
  "3",
  "3",
  "2",
  "2",
  "1",
  "3",
  "2",
  "1",
  "1",
  "2",
  "0",
  "1",
  "5",
  "3",
  "3",
  "6",
  "6",
  "1",
  "8",
  "2",
  "5",
  "8",
  "4",
  "4",
  "4",
  "5",
  "3",
  "2",
  "1",
  "1",
  "2",
  "3",
  "0",
  "1",
  "0",
  "1",
  "1",
  "1",
  "3",
  "1",
  "1",
  "4",
  "7",
  "5",
  "7",
  "12",
  "5",
  "10",
  "1",
  "3",
  "4",
  "3",
  "1",
  "0",
  "1",
  "1",
  "3",
  "1",
  "2",
  "0",
  "3",
  "2",
  "4",
  "4",
  "1",
  "4",
  "3",
  "2",
  "5",
  "3",
  "6",
  "9",
  "2",
  "2",
  "2",
  "5",
  "5",
  "1",
  "0",
  "1",
  "0",
  "0",
  "1",
  "0",
  "0",
  "4",
  "5",
  "2",
  "3",
  "4",
  "3",
  "4",
  "6",
  "5",
  "9",
  "8",
  "0",
  "4",
  "3",
  "5",
  "6",
  "2",
  "4",
  "3",
  "3",
  "0",
  "3",
  "0",
  "1",
  "3",
  "1",
  "3",
  "1",
  "2",
  "3",
  "8",
  "7",
  "4",
  "3",
  "2",
  "5",
  "3",
  "7",
  "1",
  "2",
  "1",
  "1",
  "1",
  "1",
  "4",
  "2",
  "0",
  "1",
  "2",
  "1",
  "4",
  "4",
  "4",
  "7",
  "3",
  "2",
  "7",
  "5",
  "2",
  "6",
  "4",
  "1",
]
`;

exports[`heatmap loads and renders top posts for subreddit in URL 1`] = `
Array [
  "0",
  "2",
  "1",
  "0",
  "3",
  "2",
  "2",
  "0",
  "1",
  "2",
  "1",
  "4",
  "3",
  "2",
  "2",
  "6",
  "3",
  "7",
  "7",
  "2",
  "4",
  "3",
  "3",
  "2",
  "2",
  "1",
  "3",
  "2",
  "1",
  "1",
  "2",
  "0",
  "1",
  "5",
  "3",
  "3",
  "6",
  "6",
  "1",
  "8",
  "2",
  "5",
  "8",
  "4",
  "4",
  "4",
  "5",
  "3",
  "2",
  "1",
  "1",
  "2",
  "3",
  "0",
  "1",
  "0",
  "1",
  "1",
  "1",
  "3",
  "1",
  "1",
  "4",
  "7",
  "5",
  "7",
  "12",
  "5",
  "10",
  "1",
  "3",
  "4",
  "3",
  "1",
  "0",
  "1",
  "1",
  "3",
  "1",
  "2",
  "0",
  "3",
  "2",
  "4",
  "4",
  "1",
  "4",
  "3",
  "2",
  "5",
  "3",
  "6",
  "9",
  "2",
  "2",
  "2",
  "5",
  "5",
  "1",
  "0",
  "1",
  "0",
  "0",
  "1",
  "0",
  "0",
  "4",
  "5",
  "2",
  "3",
  "4",
  "3",
  "4",
  "6",
  "5",
  "9",
  "8",
  "0",
  "4",
  "3",
  "5",
  "6",
  "2",
  "4",
  "3",
  "3",
  "0",
  "3",
  "0",
  "1",
  "3",
  "1",
  "3",
  "1",
  "2",
  "3",
  "8",
  "7",
  "4",
  "3",
  "2",
  "5",
  "3",
  "7",
  "1",
  "2",
  "1",
  "1",
  "1",
  "1",
  "4",
  "2",
  "0",
  "1",
  "2",
  "1",
  "4",
  "4",
  "4",
  "7",
  "3",
  "2",
  "7",
  "5",
  "2",
  "6",
  "4",
  "1",
]
`;
